### ISSUE TITLE
Remote shell into pods has stopped working

### ISSUE THREAD

**Issue by siamaksade** [openshift-roadshow/cloud-native-katacoda#25]
We had this issue before but was resolved. It appeared again now:
```
[root@master ~]# oc rsh inventory-4-bz6sn
Error from server (BadRequest): Upgrade request required
```

**Comment by BenHall** [comment]
I understand now why this has reappeared. The RSH target URL is based on the URL provided when the user logs in. This is what I see when I login without providing a URL. Notice, the rsh command talks directly to the master and not via the Katacoda Proxy.


```
$ oc login -u dev
Authentication required for https://master:8443 (openshift)Username: devPassword:
Login successful.
$ oc --loglevel 9 rsh docker-registry-1-zjcgt
I1215 16:33:42.371087   22215 loader.go:357] Config loaded from file /root/.kube/configI1215 16:33:42.375724   22215 round_trippers.go:386] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: oc/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff" https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgtI1215 16:33:42.546103   22215 round_trippers.go:405] GET https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgt 200 OK in 165 millisecondsI1215 16:33:42.546132   22215 round_trippers.go:411] Response Headers:
I1215 16:33:42.546139   22215 round_trippers.go:414]     Content-Type: application/jsonI1215 16:33:42.546151   22215 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:33:42 GMT
I1215 16:33:42.546157   22215 round_trippers.go:414]     Cache-Control: no-store
I1215 16:33:42.546442   22215 request.go:994] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"docker-registry-1-zjcgt","generateName":"docker-registry-1-","namespace":"default","selfLink":"/api/v1/namespaces/default/pods/docker-registry-1-zjcgt","uid":"29727e58-d75d-11e7-b76c-525400123456","resourceVersion":"5771","creationTimestamp":"2017-12-02T12:34:38Z","labels":{"deployment":"docker-registry-1","deploymentconfig":"docker-registry","docker-registry":"default"},"annotations":{"kubernetes.io/created-by":"{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"default\",\"name\":\"docker-registry-1\",\"uid\":\"25ebf81b-d75d-11e7-b76c-525400123456\",\"apiVersion\":\"v1\",\"resourceVersion\":\"2516\"}}\n","openshift.io/deployment-config.latest-version":"1","openshift.io/deployment-config.name":"docker-registry","openshift.io/deployment.name":"docker-registry-1","openshift.io/scc":"restricted"},"ownerReferences":[{"apiVersion":"v1","kind":"ReplicationController","name":"docker-registry-1","uid":"25ebf81b-d75d-11e7-b76c-525400123456","controller":true,"blockOwnerDeletion":true}]},"spec":{"volumes":[{"name":"registry-storage","emptyDir":{}},{"name":"registry-certificates","secret":{"secretName":"registry-certificates","defaultMode":420}},{"name":"registry-token-b8h8t","secret":{"secretName":"registry-token-b8h8t","defaultMode":420}}],"containers":[{"name":"registry","image":"openshift3/ose-docker-registry:v3.7.9","ports":[{"containerPort":5000,"protocol":"TCP"}],"env":[{"name":"REGISTRY_HTTP_ADDR","value":":5000"},{"name":"REGISTRY_HTTP_NET","value":"tcp"},{"name":"REGISTRY_HTTP_SECRET","value":"9yT181pYXqxvVvWGAZO7ChvGl9WjREK6Nr86wnS19hs="},{"name":"REGISTRY_MIDDLEWARE_REPOSITORY_OPENSHIFT_ENFORCEQUOTA","value":"false"},{"name":"OPENSHIFT_DEFAULT_REGISTRY","value":"docker-registry.default.svc:5000"},{"name":"REGISTRY_HTTP_TLS_KEY","value":"/etc/secrets/registry.key"},{"name":"REGISTRY_HTTP_TLS_CERTIFICATE","value":"/etc/secrets/registry.crt"}],"resources":{"requests":{"cpu":"100m","memory":"256Mi"}},"volumeMounts":[{"name":"registry-storage","mountPath":"/registry"},{"name":"registry-certificates","mountPath":"/etc/secrets"},{"name":"registry-token-b8h8t","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"livenessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTPS"},"initialDelaySeconds":10,"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"readinessProbe":{"httpGet":{"path":"/healthz","port":5000,"scheme":"HTTPS"},"timeoutSeconds":5,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"IfNotPresent","securityContext":{"capabilities":{"drop":["KILL","MKNOD","SETGID","SETUID"]},"privileged":false,"seLinuxOptions":{"level":"s0:c1,c0"},"runAsUser":1000000000}}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","nodeSelector":{"region":"infra"},"serviceAccountName":"registry","serviceAccount":"registry","nodeName":"master","securityContext":{"seLinuxOptions":{"level":"s0:c1,c0"},"fsGroup":1000000000},"imagePullSecrets":[{"name":"registry-dockercfg-s8knd"}],"schedulerName":"default-scheduler"},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-02T12:34:38Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-15T16:27:40Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-02T12:34:38Z"}],"hostIP":"172.17.0.41","podIP":"10.128.0.10","startTime":"2017-12-02T12:34:38Z","containerStatuses":[{"name":"registry","state":{"running":{"startedAt":"2017-12-15T16:19:02Z"}},"lastState":{"terminated":{"exitCode":2,"reason":"Error","startedAt":"2017-12-02T12:34:52Z","finishedAt":"2017-12-02T13:01:05Z","containerID":"docker://2c8aff1bc5f5ef268a64e4ded321fe185db49ccf9555926cb78b78d81f985619"}},"ready":true,"restartCount":1,"image":"registry.access.redhat.com/openshift3/ose-docker-registry:v3.7.9","imageID":"docker-pullable://registry.access.redhat.com/openshift3/ose-docker-registry@sha256:ebc2969d0c2b9ee3d6af7ccba9eb9bc4adba8a957e0a065e7c9714337623e2d5","containerID":"docker://ef5b89ba12197370d9ec1f43e506722b61849fb7e84e77359dcfa59e216e492b"}],"qosClass":"Burstable"}}
I1215 16:33:42.560519   22215 round_trippers.go:386] curl -k -v -XPOST  -H "X-Stream-Protocol-Version: v4.channel.k8s.io" -H "X-Stream-Protocol-Version: v3.channel.k8s.io" -H "X-Stream-Protocol-Version: v2.channel.k8s.io" -H "X-Stream-Protocol-Version: channel.k8s.io" https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgt/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=registry&container=registry&stdin=true&stdout=true&tty=true
I1215 16:33:42.560676   22215 round_trippers.go:386] curl -k -v -XPOST  -H "X-Stream-Protocol-Version: v4.channel.k8s.io" -H "X-Stream-Protocol-Version: v3.channel.k8s.io" -H "X-Stream-Protocol-Version: v2.channel.k8s.io" -H "X-Stream-Protocol-Version: channel.k8s.io" -H "User-Agent: oc/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff" https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgt/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=registry&container=registry&stdin=true&stdout=true&tty=true
I1215 16:33:42.814485   22215 round_trippers.go:405] POST https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgt/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=registry&container=registry&stdin=true&stdout=true&tty=true 101 Switching Protocols in 253 milliseconds
I1215 16:33:42.814523   22215 round_trippers.go:411] Response Headers:
I1215 16:33:42.814535   22215 round_trippers.go:414]     X-Stream-Protocol-Version: v4.channel.k8s.io
I1215 16:33:42.814544   22215 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:33:42 GMT
I1215 16:33:42.814552   22215 round_trippers.go:414]     Connection: Upgrade
I1215 16:33:42.814560   22215 round_trippers.go:414]     Upgrade: SPDY/3.1
I1215 16:33:42.814570   22215 round_trippers.go:405] POST https://master:8443/api/v1/namespaces/default/pods/docker-registry-1-zjcgt/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=registry&container=registry&stdin=true&stdout=true&tty=true 101 Switching Protocols in253 milliseconds
I1215 16:33:42.814579   22215 round_trippers.go:411] Response Headers:
I1215 16:33:42.814588   22215 round_trippers.go:414]     Connection: Upgrade
I1215 16:33:42.814596   22215 round_trippers.go:414]     Upgrade: SPDY/3.1
I1215 16:33:42.814603   22215 round_trippers.go:414]     X-Stream-Protocol-Version: v4.channel.k8s.io
I1215 16:33:42.814611   22215 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:33:42 GMT
```

When providing the URL via the login, it will use the proxy when connecting via RSH. I'm guessing out nginx config is dropping the "Connection: Upgrade / Upgrade: SPDY/3.1" header. 


```
[root@master ~]# oc --loglevel 9 rsh inventory-4-bz6sn
I1215 16:20:52.489063   50040 loader.go:357] Config loaded from file /root/.kube/config
I1215 16:20:52.493262   50040 round_trippers.go:386] curl -k -v -XGET  -H "Accept: application/json, */*" -H "User-Agent: oc/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff" -H "Authorization: Bearer 9ULLg3VuEuWyr52T84VrGL_Xr3iNV4Cb1Opptmrh7rc" https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn
I1215 16:20:52.609685   50040 round_trippers.go:405] GET https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn 200 OK in 116 milliseconds
I1215 16:20:52.610046   50040 round_trippers.go:411] Response Headers:
I1215 16:20:52.610083   50040 round_trippers.go:414]     Server: nginx/1.11.9
I1215 16:20:52.610574   50040 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:20:52 GMT
I1215 16:20:52.610604   50040 round_trippers.go:414]     Content-Type: application/json
I1215 16:20:52.610620   50040 round_trippers.go:414]     Cache-Control: no-store
I1215 16:20:52.610935   50040 request.go:994] Response Body: {"kind":"Pod","apiVersion":"v1","metadata":{"name":"inventory-4-bz6sn","generateName":"inventory-4-","namespace":"coolstore","selfLink":"/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn","uid":"5273c79f-e1b2-11e7-837d-0242ac11001d","resourceVersion":"13503","creationTimestamp":"2017-12-15T16:09:26Z","labels":{"app":"inventory","deployment":"inventory-4","deploymentconfig":"inventory","group":"com.redhat.cloudnative","provider":"fabric8","version":"1.0-SNAPSHOT"},"annotations":{"fabric8.io/iconUrl":"img/icons/wildfly.svg","fabric8.io/metrics-path":"dashboard/file/kubernetes-pods.json/?var-project=inventory\u0026var-version=1.0-SNAPSHOT","kubernetes.io/created-by":"{\"kind\":\"SerializedReference\",\"apiVersion\":\"v1\",\"reference\":{\"kind\":\"ReplicationController\",\"namespace\":\"coolstore\",\"name\":\"inventory-4\",\"uid\":\"50227b68-e1b2-11e7-837d-0242ac11001d\",\"apiVersion\":\"v1\",\"resourceVersion\":\"13452\"}}\n","openshift.io/deployment-config.latest-version":"4","openshift.io/deployment-config.name":"inventory","openshift.io/deployment.name":"inventory-4","openshift.io/scc":"restricted"},"ownerReferences":[{"apiVersion":"v1","kind":"ReplicationController","name":"inventory-4","uid":"50227b68-e1b2-11e7-837d-0242ac11001d","controller":true,"blockOwnerDeletion":true}]},"spec":{"volumes":[{"name":"volume-nvdth","configMap":{"name":"inventory","defaultMode":420}},{"name":"default-token-r4j84","secret":{"secretName":"default-token-r4j84","defaultMode":420}}],"containers":[{"name":"wildfly-swarm","image":"docker-registry.default.svc:5000/coolstore/inventory@sha256:a4a788bea61e4937d4d6bcdacd5ef9e711cd75047fa82a50d9981e20177ea06a","ports":[{"name":"http","containerPort":8080,"protocol":"TCP"},{"name":"prometheus","containerPort":9779,"protocol":"TCP"},{"name":"jolokia","containerPort":8778,"protocol":"TCP"}],"env":[{"name":"KUBERNETES_NAMESPACE","valueFrom":{"fieldRef":{"apiVersion":"v1","fieldPath":"metadata.namespace"}}},{"name":"JAVA_OPTIONS","value":"-Dswarm.project.stage=prod -Dswarm.project.stage.file=file:///app/config/project-stages.yml"}],"resources":{},"volumeMounts":[{"name":"volume-nvdth","mountPath":"/app/config"},{"name":"default-token-r4j84","readOnly":true,"mountPath":"/var/run/secrets/kubernetes.io/serviceaccount"}],"livenessProbe":{"httpGet":{"path":"/node","port":8080,"scheme":"HTTP"},"timeoutSeconds":1,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"readinessProbe":{"httpGet":{"path":"/node","port":8080,"scheme":"HTTP"},"timeoutSeconds":1,"periodSeconds":10,"successThreshold":1,"failureThreshold":3},"terminationMessagePath":"/dev/termination-log","terminationMessagePolicy":"File","imagePullPolicy":"IfNotPresent","securityContext":{"capabilities":{"drop":["KILL","MKNOD","SETGID","SETUID"]},"privileged":false,"seLinuxOptions":{"level":"s0:c10,c0"},"runAsUser":1000090000}}],"restartPolicy":"Always","terminationGracePeriodSeconds":30,"dnsPolicy":"ClusterFirst","serviceAccountName":"default","serviceAccount":"default","nodeName":"master","securityContext":{"seLinuxOptions":{"level":"s0:c10,c0"},"fsGroup":1000090000},"imagePullSecrets":[{"name":"default-dockercfg-mvrsx"}],"schedulerName":"default-scheduler"},"status":{"phase":"Running","conditions":[{"type":"Initialized","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-15T16:09:26Z"},{"type":"Ready","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-15T16:09:46Z"},{"type":"PodScheduled","status":"True","lastProbeTime":null,"lastTransitionTime":"2017-12-15T16:09:26Z"}],"hostIP":"172.17.0.29","podIP":"10.128.0.73","startTime":"2017-12-15T16:09:26Z","containerStatuses":[{"name":"wildfly-swarm","state":{"running":{"startedAt":"2017-12-15T16:09:28Z"}},"lastState":{},"ready":true,"restartCount":0,"image":"docker-registry.default.svc:5000/coolstore/inventory:latest","imageID":"docker-pullable://docker-registry.default.svc:5000/coolstore/inventory@sha256:a4a788bea61e4937d4d6bcdacd5ef9e711cd75047fa82a50d9981e20177ea06a","containerID":"docker://e0a4e8fbc3cdb5641eb45deabc13440d075baa2013c0b0709d32cc7e81666454"}],"qosClass":"BestEffort"}}
I1215 16:20:52.613653   50040 round_trippers.go:386] curl -k -v -XPOST  -H "X-Stream-Protocol-Version: v4.channel.k8s.io" -H "X-Stream-Protocol-Version: v3.channel.k8s.io" -H "X-Stream-Protocol-Version: v2.channel.k8s.io" -H "X-Stream-Protocol-Version: channel.k8s.io" https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=wildfly-swarm&container=wildfly-swarm&stdin=true&stdout=true&tty=true
I1215 16:20:52.613694   50040 round_trippers.go:386] curl -k -v -XPOST  -H "X-Stream-Protocol-Version: v4.channel.k8s.io" -H "X-Stream-Protocol-Version: v3.channel.k8s.io" -H "X-Stream-Protocol-Version: v2.channel.k8s.io" -H "X-Stream-Protocol-Version: channel.k8s.io" -H "User-Agent: oc/v1.7.6+a08f5eeb62 (linux/amd64) kubernetes/c84beff" -H "Authorization: Bearer 9ULLg3VuEuWyr52T84VrGL_Xr3iNV4Cb1Opptmrh7rc" https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=wildfly-swarm&container=wildfly-swarm&stdin=true&stdout=true&tty=true
I1215 16:20:52.713423   50040 round_trippers.go:405] POST https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=wildfly-swarm&container=wildfly-swarm&stdin=true&stdout=true&tty=true 400 Bad Request in 99 milliseconds
I1215 16:20:52.713457   50040 round_trippers.go:411] Response Headers:
I1215 16:20:52.713469   50040 round_trippers.go:414]     Server: nginx/1.11.9
I1215 16:20:52.713477   50040 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:20:52 GMT
I1215 16:20:52.713485   50040 round_trippers.go:414]     Content-Type: application/json
I1215 16:20:52.713492   50040 round_trippers.go:414]     Content-Length: 139
I1215 16:20:52.713499   50040 round_trippers.go:414]     Connection: keep-alive
I1215 16:20:52.713505   50040 round_trippers.go:414]     Cache-Control: no-store
I1215 16:20:52.713514   50040 round_trippers.go:405] POST https://2886795293-8443-simba02.environments.katacoda.com:443/api/v1/namespaces/coolstore/pods/inventory-4-bz6sn/exec?command=%2Fbin%2Fsh&command=-c&command=TERM%3D%22xterm%22+%2Fbin%2Fsh&container=wildfly-swarm&container=wildfly-swarm&stdin=true&stdout=true&tty=true 400 Bad Request in 99 milliseconds
I1215 16:20:52.713526   50040 round_trippers.go:411] Response Headers:
I1215 16:20:52.713534   50040 round_trippers.go:414]     Connection: keep-alive
I1215 16:20:52.713541   50040 round_trippers.go:414]     Cache-Control: no-store
I1215 16:20:52.713548   50040 round_trippers.go:414]     Server: nginx/1.11.9
I1215 16:20:52.713556   50040 round_trippers.go:414]     Date: Fri, 15 Dec 2017 16:20:52 GMT
I1215 16:20:52.713563   50040 round_trippers.go:414]     Content-Type: application/json
I1215 16:20:52.713570   50040 round_trippers.go:414]     Content-Length: 139
I1215 16:20:52.714258   50040 helpers.go:206] server response object: [{
"kind": "Status",
"apiVersion": "v1",
"metadata": {},
"status": "Failure",
"message": "Upgrade request required",
"reason": "BadRequest",
"code": 400
}]
F1215 16:20:52.714285   50040 helpers.go:120] Error from server (BadRequest): Upgrade request required
```

Fix:
1) With the oAuth Config fixed, it doesn't be required to provide the URL (hopefully, given the Jenkins bug).

2) We'll fix NGINX/Proxy config not to drop the headers.

**Comment by BenHall** [comment]
It's something to do with the proxy, but still trying to work out what.

Temp workaround :(

`oc --server https://master:8443 rsh`

**Comment by BenHall** [comment]
"After moving to version 1.9.5, you can no longer configure NGINX to use SPDY. If you would like to keep using SPDY, it will continue to be supported in the NGINX 1.8.x branch."  https://www.nginx.com/blog/nginx-1-9-5/

**Comment by BenHall** [comment]
This is proving problematic. It looks like SPDY isn't supported by most load balancers. Enabling it via a custom build of nginx caused requests to stop responding. I looked at changing LB but HAProxy, Envoy and Traefik don't appear to support it.

In the meantime, I've set the workaround on the scenario commands - https://github.com/openshift-roadshow/cloud-native-katacoda/commit/b51822c40f227ff491138d6b99958cf9b57dd96a
